modules/subschema/src/main/scala/com.snowplowanalytics/iglu.schemaddl/jsonschema/subschema/package.scala

```diff
@@ -54,83 +63,109 @@ package object subschema {
       .getOrElse(s)
   }
 
+  @tailrec
+  def canonicalizeObject(s: Schema): Schema =
+    (s.properties, s.additionalProperties, s.patternProperties) match {
+      case (_, Some(AdditionalPropertiesAllowed(allowed)), _) =>
+        canonicalizeObject(s.copy(additionalProperties = Some(AdditionalPropertiesSchema(if (allowed) any else none))))
+      case (Some(Properties(props)), Some(AdditionalPropertiesSchema(_)), pProps) =>
+        val newPatternProps = props.map({case (k, v) => (s"^$k$$", v)}) ++ pProps.map(_.value).getOrElse(Map.empty)
+        // TODO: rewrite additionalProperties into patternProperties
+        s.copy(properties = None, patternProperties = Some(PatternProperties(newPatternProps)))
+      // TODO: handle overlapping patternProperties
+      case _ => s
+    }
```

Comment:
The overlapping could be challenging, do you think we should create a new task for that? 

I can see in the paper a reference to `dependencies`.
![Screenshot 2023-03-06 at 11 24 26](https://user-images.githubusercontent.com/6108743/223083823-289c226d-2deb-4ad7-a2e2-589817424b31.png). That should be transformed to `required` but I can't see that node in the AST. 🤔

Reply:
Yeah I'll create a ticket for overlapping pattern properties 👍🏻

Reply:
https://snplow.atlassian.net/browse/XP-1365